### PR TITLE
[mlir] Add config for PDL

### PR DESCRIPTION
**Pull request by jpienaar**
Make it so that PDL in pattern rewrites can be optionally disabled.

PDL is still enabled by default and not optional bazel. So this should be a NOP for most folks, while enabling other to disable.

This is piped through mlir-tblgen invocation and that could be changed/avoided by splitting up the passes file instead.

This only works with tests disabled. With tests enabled this still compiles but tests fail as there is no lit config to disable tests that depend on PDL rewrites yet.

### CODE REVIEW

**Comment by joker-eph** [comment]
Can you expand on the motivation? Why is it a problem that PDL is always included? Your description isn’t very explicit on the impact.

**Comment by jpienaar** [comment]
> Can you expand on the motivation? Why is it a problem that PDL is always included? Your description isn’t very explicit on the impact.

Done. Its not PDL specific, I think it is a problem if any dialect is always included even if not used :) The others just have a simple method to elide.

**Comment by joker-eph** [comment]
Need to iterate a bit more on framing the problem:

> MLIR is a general infrastructure and as far as possible no dialect is intended to be special or privileged

Yes, and you can use MLIR without PDL, it's not a mandatory component. I believe that the minimal examples `mlir-cat` and `mlir-minimal-opt` don't have PDL linked in (so you can already have your *-opt tool without it!

> Users should be able to use MLIR and the parts they need to customize for their solution (e.g., one doesn't have to include any dialect except the ones one uses). 

I don't see PDL as "just a dialect": I see it more as an infrastructure component. 
That is no one will have PDL used within their IR! 
It isn't comparable to any other dialect and completely unique from this point of view (do we have other dialects that aren't targeting user-written compilers? The transform dialect might be in-between)

> Many parts are elided when not referenced (LTO DCE'd etc), but this is not possible with PDL given how its integrated with the common rewrite drivers. This results in it always being included even when not used.

Now that seems more accurate to me: the dependency on PDL is only from the rewrite drivers, which is a "bring your own" thing by the way.
So the underlying question for motivating this here would be better framed IMO as "should we have an option to build the GreedyPatternRewriter without PDL?" 
I see it less as a strong need here (because again, bring your own driver if you don't like it).

Without a stronger case, this seems like something I would be supportive if we can make it minimally intrusive: that is localize the changes to a maximum and not spread through the codebase.
There are far too many #ifdef  to me right now, this may need some more refactoring first to avoid this and make PDL a more proper "separate component" inside libMLIRRewrite so that is can be enabled/disabled more naturally.

File: mlir/include/mlir/IR/PDLPatternMatch.h.inc [hunk 1]
```diff
@@ -0,0 +1,992 @@
+//===- PDLPatternMatch.h - PDLPatternMatcher classes -------==---*- C++ -*-===//
```
Comment (joker-eph) [inline mlir/include/mlir/IR/PDLPatternMatch.h.inc:1]: Not clear to me why is this a `.h.inc` extension? Seems like a regular `.h` to me.

Isn't it standalone? If so can you point out why?

File: mlir/include/mlir/IR/PDLPatternMatch.h.inc [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_IR_PDLPATTERNMATCH_H
+#define MLIR_IR_PDLPATTERNMATCH_H
+
+#if MLIR_ENABLE_PDL
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
```
Comment (joker-eph) [inline mlir/include/mlir/IR/PDLPatternMatch.h.inc:14]: There is no include from the to the PDL dialect: where is the coupling?

File: mlir/include/mlir/IR/PatternMatch.h [hunk 3]
```diff
@@ -1863,6 +944,8 @@ class RewritePatternSet {
 
   MLIRContext *const context;
   NativePatternListT nativePatterns;
+
+  // Technically not needed if PDL not enabled. Stubs enable less ifdefs.
```
Comment (joker-eph) [inline mlir/include/mlir/IR/PatternMatch.h:948]:
```suggestion
  /// Patterns expressed with PDL. This will compile to a stub class when PDL is not enabled.
```

File: mlir/include/mlir/InitAllDialects.h [hunk 4]
```diff
@@ -14,6 +14,7 @@
 #ifndef MLIR_INITALLDIALECTS_H_
 #define MLIR_INITALLDIALECTS_H_
 
+#include "mlir/Config/mlir-config.h"
```
Comment (joker-eph) [inline mlir/include/mlir/InitAllDialects.h:17]: Why is it needed here?

File: mlir/include/mlir/InitAllExtensions.h [hunk 5]
```diff
@@ -14,6 +14,7 @@
 #ifndef MLIR_INITALLEXTENSIONS_H_
 #define MLIR_INITALLEXTENSIONS_H_
 
+#include "mlir/Config/mlir-config.h"
```
Comment (joker-eph) [inline mlir/include/mlir/InitAllExtensions.h:17]: Same question?

File: mlir/include/mlir/InitAllExtensions.h [hunk 6]
```diff
@@ -76,6 +76,7 @@ inline void registerAllExtensions(DialectRegistry &registry) {
   sparse_tensor::registerTransformDialectExtension(registry);
   tensor::registerTransformDialectExtension(registry);
   transform::registerLoopExtension(registry);
+  bufferization::registerTransformDialectExtension(registry);
```
Comment (joker-eph) [inline mlir/include/mlir/InitAllExtensions.h:79]: Does not seem to belong to this diff

File: mlir/include/mlir/Rewrite/PatternApplicator.h [hunk 7]
```diff
@@ -92,6 +92,7 @@ class PatternApplicator {
   /// The set of patterns that may match against any operation type, stable
   /// sorted by benefit.
   SmallVector<const RewritePattern *, 1> anyOpPatterns;
+
```
Comment (joker-eph) [inline mlir/include/mlir/Rewrite/PatternApplicator.h:95]: Please remove: this is the only change in this file.

File: mlir/lib/Rewrite/CMakeLists.txt [hunk 8]
```diff
@@ -1,3 +1,14 @@
...
+
+if(MLIR_ENABLE_PDL)
+set(MLIRRewritePDLDeps
+  MLIRPDLDialect
+  MLIRPDLInterpDialect
+  MLIRPDLToPDLInterp)
+else()
+  set(MLIRRewritePDLDeps)
+endif()
+
```
Comment (joker-eph) [inline mlir/lib/Rewrite/CMakeLists.txt:11]: Can we also handle `ByteCode.cpp` here instead of adding `#ifdef` in the code?

It could be its own library and MLIRRewrite could depend on it:

```
if(MLIR_ENABLE_PDL)
  add_mlir_library(MLIRRewriteBytecode
    ByteCode.cpp
  DEPENDS
    MLIRPDLDialect
    MLIRPDLInterpDialect
    MLIRPDLToPDLInterp)
else()
  add_mlir_library(MLIRRewriteBytecode)
endif()
```

File: mlir/lib/Transforms/Utils/DialectConversion.cpp [hunk 9]
```diff
@@ -3312,6 +3312,7 @@ auto ConversionTarget::getOpInfo(OperationName op) const
   return std::nullopt;
 }
 
+#if MLIR_ENABLE_PDL
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/DialectConversion.cpp:3315]: Can we extract all of this in a separate .cpp and handle this with separate CMakeLibraries?

File: mlir/python/CMakeLists.txt [hunk 10]
```diff
@@ -472,18 +472,18 @@ declare_mlir_python_extension(MLIRPythonExtension.Dialects.Quant.Pybind
...
+      DialectPDL.cpp
+    PRIVATE_LINK_LIBS
+      LLVMSupport
+    EMBED_CAPI_LINK_LIBS
+      MLIRCAPIIR
+      MLIRCAPIPDL
+  )
 
 declare_mlir_python_extension(MLIRPythonExtension.Dialects.SparseTensor.Pybind
   MODULE_NAME _mlirDialectsSparseTensor
```
Comment (joker-eph) [inline mlir/python/CMakeLists.txt:489]: This entire file is formatting change now, can we revert this?

File: mlir/tools/mlir-opt/mlir-opt.cpp [hunk 11]
```diff
@@ -253,12 +252,17 @@ void registerTestPasses() {
...
   mlir::test::registerTestVectorReductionToSPIRVDotProd();
   mlir::test::registerTestNvgpuLowerings();
   mlir::test::registerTestWrittenToPass();
+#if MLIR_ENABLE_PDL
+  mlir::test::registerTestDialectConversionPasses();
+#endif
+  mlir::test::registerTestPDLByteCodePass();
+  mlir::test::registerTestPDLLPasses();
+  mlir::test::registerTestTransformDialectEraseSchedulePass();
+  mlir::test::registerTestTransformDialectInterpreterPass();
```
Comment (joker-eph) [inline mlir/tools/mlir-opt/mlir-opt.cpp:265]: Why did you move these all? They don't seem covered by the `#if` anyway.

File: mlir/lib/IR/PatternMatch.cpp [hunk 12]
```diff
@@ -97,6 +98,7 @@ LogicalResult RewritePattern::match(Operation *op) const {
 /// Out-of-line vtable anchor.
 void RewritePattern::anchor() {}
 
+#if MLIR_ENABLE_PDL
```
Comment (joker-eph) [inline mlir/lib/IR/PatternMatch.cpp:101]: Can we extract the code in this file into a new .cpp and make it its own CMake library to manage enabling/disabling?

File: mlir/lib/Rewrite/ByteCode.cpp [hunk 13]
```diff
@@ -2333,3 +2334,4 @@ LogicalResult PDLByteCode::rewrite(PatternRewriter &rewriter,
   }
   return result;
 }
+#endif // MLIR_ENABLE_PDL
```
Comment (joker-eph) [inline mlir/lib/Rewrite/ByteCode.cpp:2337]: I think we can avoid the macros here by handling this as a library in cmake.

**Review by joker-eph — COMMENTED** [review]
Look pretty good to me overall! 

> MLIR is a general infrastructure and as far as possible no dialect is intended to be special or privileged (we have Builtin which is both and should be addressed ...). Users should be able to use MLIR and the parts they need to customize for their solution (e.g., one doesn't have to include any dialect except the ones one uses). 

Can you rework the description please? I feel it is misleading.
PDL/PDLL are dialects for the pattern rewrite infra, and what you're doing is making is possible for the "pattern rewrite infrastructure" to statically disable support for PD/PDLL. This does not make any dialect more "privileged", or "special". The fact that these are "dialects" is anecdotical: we could have written a different DSL with a custom bytecode interpret without using MLIR dialect just as well. The misleading part is that this is describing this in the say way as if these were dialects like the others, intended to blend in the compiler ecosystem.

File: mlir/include/mlir/InitAllDialects.h (hunk 4 above)
Comment (jpienaar) [reply mlir/include/mlir/InitAllDialects.h:17]: Was needed when PDL was optionally built, removed.

File: mlir/lib/Transforms/Utils/DialectConversion.cpp (hunk 9 above)
Comment (jpienaar) [reply mlir/lib/Transforms/Utils/DialectConversion.cpp:3315]: This one is trickier as the parts here depend on ConversionPatternRewriterImpl, so would result in weird library split.

File: mlir/tools/mlir-opt/mlir-opt.cpp (hunk 11 above)
Comment (jpienaar) [reply mlir/tools/mlir-opt/mlir-opt.cpp:265]: Returned. These were previously covered yes/some may have been a fake dep as Matthias pointed out.

File: mlir/include/mlir/IR/PDLPatternMatch.h.inc (hunk 1 above)
Comment (joker-eph) [reply mlir/include/mlir/IR/PDLPatternMatch.h.inc:1]: ping?

File: mlir/lib/Transforms/Utils/CMakeLists.txt [hunk 14]
```diff
@@ -23,3 +23,5 @@ add_mlir_library(MLIRTransformUtils
   MLIRSubsetOpInterface
   MLIRRewrite
   )
+
+
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/CMakeLists.txt:27]: Nite: Remove, no other changes in this file.

File: mlir/lib/Tools/CMakeLists.txt [hunk 15]
```diff
@@ -1,11 +1,11 @@
 add_subdirectory(lsp-server-support)
 add_subdirectory(mlir-lsp-server)
 add_subdirectory(mlir-opt)
-add_subdirectory(mlir-pdll-lsp-server)
 add_subdirectory(mlir-query)
 add_subdirectory(mlir-reduce)
 add_subdirectory(mlir-tblgen)
 add_subdirectory(mlir-translate)
 add_subdirectory(PDLL)
+add_subdirectory(mlir-pdll-lsp-server)
```
Comment (joker-eph) [inline mlir/lib/Tools/CMakeLists.txt:9]: Revert: no changes other than this in this file.

**Comment by joker-eph** [comment]
> Look pretty good to me overall!
> 
> > MLIR is a general infrastructure and as far as possible no dialect is intended to be special or privileged (we have Builtin which is both and should be addressed ...). Users should be able to use MLIR and the parts they need to customize for their solution (e.g., one doesn't have to include any dialect except the ones one uses).
> 
> Can you rework the description please? I feel it is misleading. PDL/PDLL are dialects for the pattern rewrite infra, and what you're doing is making is possible for the "pattern rewrite infrastructure" to statically disable support for PD/PDLL. This does not make any dialect more "privileged", or "special". The fact that these are "dialects" is anecdotical: we could have written a different DSL with a custom bytecode interpret without using MLIR dialect just as well. The misleading part is that this is describing this in the say way as if these were dialects like the others, intended to blend in the compiler ecosystem.

Ping :)

**Comment by jpienaar** [comment]
> > Look pretty good to me overall!
> > > MLIR is a general infrastructure and as far as possible no dialect is intended to be special or privileged (we have Builtin which is both and should be addressed ...). Users should be able to use MLIR and the parts they need to customize for their solution (e.g., one doesn't have to include any dialect except the ones one uses).
> > 
> > 
> > Can you rework the description please? I feel it is misleading. PDL/PDLL are dialects for the pattern rewrite infra, and what you're doing is making is possible for the "pattern rewrite infrastructure" to statically disable support for PD/PDLL. This does not make any dialect more "privileged", or "special". The fact that these are "dialects" is anecdotical: we could have written a different DSL with a custom bytecode interpret without using MLIR dialect just as well. The misleading part is that this is describing this in the say way as if these were dialects like the others, intended to blend in the compiler ecosystem.
> 
> Ping :)

Done, simplified it to just that.

File: mlir/include/mlir/IR/PDLPatternMatch.h.inc (hunk 1 above)
Comment (jpienaar) [reply mlir/include/mlir/IR/PDLPatternMatch.h.inc:1]: Its not as it depends on data structures defined above it, but also it doesn't make sense to use in isolation as the functionality is not self-contained.

File: mlir/include/mlir/IR/PDLPatternMatch.h.inc (hunk 2 above)
Comment (jpienaar) [reply mlir/include/mlir/IR/PDLPatternMatch.h.inc:16]: These are structures that only make sense with PDL, I want these all to be clearly stubbed out so that it returns empty arrays etc. Could probably leave in and hope DCE does its thing during compilation, but wanted to be more explicit.

**Review by joker-eph — APPROVED** [review]
